astroscrappy/astroscrappy.pyx

```diff
@@ -248,18 +285,43 @@ def detect_cosmics(indat, inmask=None, float sigclip=4.5, float sigfrac=0.3,
         s = rebin(conved)
         del conved
 
-        # Build a the noise map, to compare the laplacian to
+        # Build a noise map, to compare the laplacian to, based on the input
+        # data or (if supplied) input variance. If variance is used, it still
+        # gets filtered, since we can best identify cosmic rays using what the
+        # noise in a pixel *should* be, rather than what it actually is after
+        # a cosmic ray hit.
         if sepmed:
-            m5 = sepmedfilt7(cleanarr)
+            m5 = sepmedfilt7(cleanvar)
         else:
-            m5 = medfilt5(cleanarr)
+            m5 = medfilt5(cleanvar)
+
+        if var is None:
+            # Add any object/sky model flux back to the background estimate in
+            # order to estimate the noise accurately (first saving a copy if
+            # using the same median to replace bad values):
+            if cleantype == 'median':
+                m5_nobkg = m5 if bkg is None else m5.copy()
+            if bkg is not None:
+                m5 += (bkg * gain)
+        else:
+            # Median filter the main data array only if it's used for
+            # estimating cleaned values. The variance doesn't have background
+            # subtracted, so no need to add it here.
+            if cleantype == 'median':
+                if sepmed:
+                    m5_nobkg = sepmedfilt7(cleanarr)
+                else:
+                    m5_nobkg = medfilt5(cleanarr)
 
         # Clip noise so that we can take a square root
         m5[m5 < 0.00001] = 0.00001
-        noise = np.sqrt(m5 + readnoise * readnoise)
-
-        if cleantype != 'median':
+        if var is None:
+            noise = np.sqrt(m5 + readnoise * readnoise)
```

Comment:
Instead of using m5 in the noise, why don't you use the 5x5 median of cleanvar. Basically calculate one m5 for the cleanarr for the cr detection and one m5 for the noise calculation. Then you don't have to worry about adding back in bkg etc.